Source/Core/Core/IPC_HLE/WII_IPC_HLE.cpp

```diff
@@ -365,93 +365,107 @@ static std::shared_ptr<IWII_IPC_HLE_Device> GetUnusedESDevice()
+  else if (request.path.find("/dev/") == 0)
   {
-    device = GetDeviceByName(device_name);
+    device = GetDeviceByName(request.path);
   }
-  else if (device_name.find('/') == 0)
+  else if (request.path.find('/') == 0)
   {
-    device = std::make_shared<CWII_IPC_HLE_Device_FileIO>(new_fd, device_name);
+    device = std::make_shared<CWII_IPC_HLE_Device_FileIO>(new_fd, request.path);
   }
 
   if (!device)
   {
-    ERROR_LOG(WII_IPC_HLE, "Unknown device: %s", device_name.c_str());
+    ERROR_LOG(WII_IPC_HLE, "Unknown device: %s", request.path.c_str());
     return IPC_ENOENT;
   }
 
-  Memory::Write_U32(new_fd, address + 4);
-  device->Open(address, open_mode);
-  const s32 open_return_code = Memory::Read_U32(address + 4);
-  if (open_return_code < 0)
-    return open_return_code;
+  const IOSReturnCode code = device->Open(request);
+  if (code < IPC_SUCCESS)
+    return code;
   s_fdmap[new_fd] = device;
   return new_fd;
 }
 
-static IPCCommandResult HandleCommand(const u32 address)
+static IPCCommandResult HandleCommand(IOSResourceRequest& request)
 {
-  const auto command = static_cast<IPCCommandType>(Memory::Read_U32(address));
-  if (command == IPC_CMD_OPEN)
+  if (request.command == IPC_CMD_OPEN)
   {
-    const s32 new_fd = OpenDevice(address);
-    Memory::Write_U32(new_fd, address + 4);
+    IOSResourceOpenRequest open_request{request.address};
+    const s32 new_fd = OpenDevice(open_request);
+    open_request.SetReturnValue(new_fd);
     return IWII_IPC_HLE_Device::GetDefaultReply();
   }
 
-  const s32 fd = Memory::Read_U32(address + 8);
-  const auto device = (fd >= 0 && fd < IPC_MAX_FDS) ? s_fdmap[fd] : nullptr;
+  const auto device = (request.fd < IPC_MAX_FDS) ? s_fdmap[request.fd] : nullptr;
   if (!device)
   {
-    Memory::Write_U32(IPC_EINVAL, address + 4);
+    request.SetReturnValue(IPC_EINVAL);
     return IWII_IPC_HLE_Device::GetDefaultReply();
   }
 
-  switch (command)
+  switch (request.command)
   {
   case IPC_CMD_CLOSE:
-    s_fdmap[fd].reset();
-    // A close on a valid device returns IPC_SUCCESS.
-    Memory::Write_U32(IPC_SUCCESS, address + 4);
-    return device->Close(address);
+  {
+    s_fdmap[request.fd].reset();
+    device->Close();
+    request.SetReturnValue(IPC_SUCCESS);
+    return IWII_IPC_HLE_Device::GetDefaultReply();
+  }
   case IPC_CMD_READ:
-    return device->Read(address);
+  {
+    IOSResourceReadWriteRequest read_request{request.address};
+    return device->Read(read_request);
+  }
   case IPC_CMD_WRITE:
-    return device->Write(address);
+  {
+    IOSResourceReadWriteRequest write_request{request.address};
+    return device->Write(write_request);
+  }
   case IPC_CMD_SEEK:
-    return device->Seek(address);
+  {
+    IOSResourceSeekRequest seek_request{request.address};
+    return device->Seek(seek_request);
+  }
   case IPC_CMD_IOCTL:
-    return device->IOCtl(address);
+  {
+    IOSResourceIOCtlRequest ioctl_request{request.address};
+    ioctl_request.ClearOutputBuffer();
```

Comment:
That would make sense. I guess it's unsafe to do this automatically, even for the output buffer in ioctls… Removed this, and only added it back in cases where clearing the buffer is really needed.